lib/Sema/CSDiagnostics.cpp

```diff
@@ -110,6 +110,67 @@ Optional<SelectedOverload> FailureDiagnostic::getChoiceFor(Expr *expr) const {
   return getOverloadChoiceIfAvailable(cs.getCalleeLocator(expr));
 }
 
+Optional<FunctionArgApplyInfo>
+FailureDiagnostic::getFunctionArgApplyInfo() const {
+  auto &cs = getConstraintSystem();
+  auto *anchor = getRawAnchor();
+  auto path = getLocator()->getPath();
+
+  // Look for the apply-arg-to-param element in the locator's path. We may
+  // have to look through other elements that are generated from an argument
+  // conversion such as GenericArgument for an optional-to-optional conversion,
+  // and OptionalPayload for a value-to-optional conversion.
+  auto applyArgElt =
+      std::find_if(path.rbegin(), path.rend(), [](LocatorPathElt elt) {
+        return elt.getKind() == ConstraintLocator::ApplyArgToParam;
+      });
+
+  if (applyArgElt == path.rend())
+    return None;
+
+  assert(std::find_if(applyArgElt + 1, path.rend(), [](LocatorPathElt elt) {
+    return elt.getKind() == ConstraintLocator::ApplyArgToParam;
+  }) == path.rend() && "Multiple ApplyArgToParam components?");
+
+  // Form a new locator that ends at the apply-arg-to-param element, and
+  // simplify it to get the full argument expression.
+  auto argPath = path.drop_back(applyArgElt - path.rbegin());
+  auto *argLocator = cs.getConstraintLocator(
+      anchor, argPath, ConstraintLocator::getSummaryFlagsForPath(argPath));
+
+  auto *argExpr = simplifyLocatorToAnchor(cs, argLocator);
+
+  // If we were unable to simplify down to the argument expression, we don't
+  // know what this is.
+  if (!argExpr)
+    return None;
+
+  ValueDecl *callee = nullptr;
+  AnyFunctionType *fnType = nullptr;
+  if (auto overload = getChoiceFor(anchor)) {
+    // If we have resolved an overload for the callee, then use that to get the
+    // function type and callee.
+    if (auto *decl = overload->choice.getDeclOrNull())
+      callee = decl;
+
+    fnType = resolveType(overload->openedType)->getAs<AnyFunctionType>();
+  } else {
```

Comment:
I think we should always get an overload, if we didn't then wrong locator was used? Because function applied to arguments supposed to be resolved down to declaration...

Reply:
@xedin Not if it's an arbitrary function expression such as `{ y in ... }(x)`.

Reply:
Ah right, I keep on forgetting that we allow that... It's kind of bad though because there should be a way to distinguish whether declaration should be there or not, might be worse adding an assert in there or something...

Reply:
@xedin Did you mean "might be worth adding an assert"? If so, did you have a specific condition in mind? The best I can think of is making sure the apply's function expr isn't a `DeclRefExpr`, `MemberRefExpr`, `UnresolvedDotExpr`, `UnresolvedMemberExpr`, etc. in the `else` branch. Though that's pretty suboptimal as it'll be easy to miss new cases as they arise.

Reply:
Yeah, sorry I meant "worth" :) We can go with what you suggest or I think it might be sufficient to check that function expr is a `ClosureExpr` and see if we could catch some more cases when declaration couldn't be found.

Reply:
@xedin Unfortunately I believe this can also arise for various things other than `ClosureExpr`, e.g `({ y in ... }, "").0(x)`, `curriedFn()(x)`, `[{y in ... }][0](x)` to name a few.

Reply:
I think all of those supposed to produce an "overload choice" at least e.g. `.0(x)` and `[0](x)` curried cases would as well, so maybe instead of using decl we should store OverloadChoice in new structure.

Reply:
@xedin Ah yes, you're right about `.0(x)`, I forgot we have overload choices for tuple elements!

For the others though, we shouldn't get overload choices from their "callee locators". This is especially important for the curried cases, as that could lead us to looking at the wrong parameter list. Actually I've just noticed a bug in `getCalleeLocator` that means we incorrectly return a subscript overload choice for the following:

```swift
struct S {
  subscript<T>() -> (T, T) -> Void { { _, _ in } }
}

func foo(_ x: @escaping () -> Int, _ y: () -> Int) {
  let s = S()
  _ = s[]({0}, y)
}
```

which leads to a crash because we're trying to get the second parameter from the subscript decl. What should've happened is we shouldn't have resolved a callee, and we should have worked with the function type of the apply instead.

Reply:
Yeah, that makes sense. Looks like curried calls as well as literal closure calls are good candidates for such check.